metricflow/model/dbt_transformations/dbt_metric_to_dimensions_rules.py

```diff
@@ -0,0 +1,125 @@
+from datetime import datetime
+import traceback
+from typing import List, Tuple
+
+from dbt_metadata_client.dbt_metadata_api_schema import MetricNode, MetricFilter
+from metricflow.model.dbt_transformations.dbt_transform_rule import (
+    DbtTransformRule,
+    DbtTransformedObjects,
+    assert_metric_model_name,
+)
+from metricflow.model.objects.elements.dimension import Dimension, DimensionType, DimensionTypeParams
+from metricflow.time.time_granularity import TimeGranularity
+from metricflow.model.validations.validator_helpers import ModelValidationResults, ValidationIssue, ValidationError
+
+
+class DbtDimensionsToDimensions(DbtTransformRule):
+    """Rule for mapping dbt metric dimensions to data source dimensions"""
+
+    @staticmethod
+    def run(dbt_metrics: Tuple[MetricNode, ...], objects: DbtTransformedObjects) -> ModelValidationResults:  # noqa: D
+        issues: List[ValidationIssue] = []
+        for metric in dbt_metrics:
+            # Skip metrics which don't have dimensions or a model to attach them to
+            if metric.dimensions and len(metric.dimensions) > 0 and metric.model:
+                try:
+                    assert_metric_model_name(metric=metric)
+                    for dimension in metric.dimensions:
+                        objects.dimensions[metric.model.name][dimension] = Dimension(
+                            name=dimension,
+                            type=DimensionType.CATEGORICAL,
```

Comment:
Yes, and thus what about to explain is real gnarly 🙃 Sooooo currently our solution is to run the `DbtTimestampToDimension` rule after this rule. Thus any any dimension that is either in both `dbt_metric.dimensions` and `dbt_metric.timestamp` or just in `dbt_metric.timestamp` will get qualified as a `TIME` type dimension. The gnarly part is that this breaks if the run order of `DbtDimensionsToDimensions` and `DbtTimestampToDimension`. Which is also not documented anywhere in this PR until now 🙈

However, there is a `dbt_metric.model.columns` attribute. Conveniently each column has a `name`, `type`, and a few other attributes.  The issue previously with `columns` was that we didn't know if the `type` value was data warehouse dependent or dbt defined. Recently though I asked dbt people about this and, unless I misunderstood, these types are dbt defined. This means that we can update this rule to check any name in `dbt_metric.dimensions` against the columns.

As another alternative we could use the new information we have about `dbt_metric.model.columns` to build all of the dimensions from them directly instead of using `dbt_metric.dimensions`, `dbt_metric.timestamp`, and `dbt_metric.filters`. Since dbt doesn't currently support identifiers, evert column is either a dimension (either `CATEGORICAL` or `TIME`) or a measure. The issue with this approach is that rules shouldn't be dependent on other rules or the order of rules. If we were to build this rule that operates directly on columns, it wouldn't know what should be a measure (I believe). And thus the rule for building measures from metrics would have to come after this new rule and also remove dimensions from the result `objects.dimensions` that should be measures. We could potentially get around this edge case by altering the generated measure name in a deterministic way, but at that point I think we're adding more confusion than necessary as well 🙃

In conclusion, I think the solution of just modifying the `DbtDimensionsToDimensions` is the sanest.

Reply:
Resolved by: https://github.com/transform-data/metricflow/pull/324/commits/63fd39b2daa0051366b4a3471dc7c48dce652a58 & https://github.com/transform-data/metricflow/pull/324/commits/83bb37cfb7d0e3b1f9c4a4e6368b89509f65e4f2